nixos/gdm: add xorg option

getchoo opened the pull request:
Adds a new ~~override to `gnome-session` and a~~ user friendly toggle in the GDM module for removing GNOME's X11 session

Work towards https://github.com/NixOS/nixpkgs/issues/382463

<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [x] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

getchoo committed:
nixos/tests/gnome: disable xorg in gdm

getchoo commented:
And as said in the issue:

> some other distributions (like Fedora) are disabling the X11 sessions out of the box.

Not sure if we'd want to follow suit here for 25.05 or 25.11, but something to keep in mind :p

jtojnar commented on nixos/modules/services/x11/desktop-managers/gnome.nix:
```diff
@@ -189,6 +189,12 @@ in
+      waylandOnly = lib.mkOption {
+        type = lib.types.bool;
+        default = false;
+        description = "Whether to exclusively use the Wayland session.";
```
Maybe we should manage expectations that this will still keep X11 dependencies in most packages?

jtojnar commented on nixos/modules/services/x11/desktop-managers/gnome.nix:
```diff
@@ -215,7 +221,9 @@ in
 
-      services.displayManager.sessionPackages = [ pkgs.gnome-session.sessions ];
+      services.displayManager.sessionPackages = [
+        (pkgs.gnome-session.override { x11Support = !cfg.waylandOnly; }).sessions
```
Lower, we use `gnome-session` for portal configs. Same in the `gdm` module. I wonder if we should use the overridden packages there as well.

Probably not necessary, since we only take the `sessions` output here.

jtojnar commented on nixos/modules/services/x11/desktop-managers/gnome.nix:
```diff
@@ -215,7 +221,9 @@ in
 
-      services.displayManager.sessionPackages = [ pkgs.gnome-session.sessions ];
+      services.displayManager.sessionPackages = [
+        (pkgs.gnome-session.override { x11Support = !cfg.waylandOnly; }).sessions
```
Also not sure if inclusion in a module is enough for Hydra to put it into cache. But maybe NixOS tests are enough?

getchoo replied:
We could enable the option [here](https://github.com/NixOS/nixpkgs/blob/a466f5f18289cb29f58f0459f362132b34af254b/nixos) in the release jobset. Given that we are technically using both packages as you explained above, this should ensure both are always cached

jtojnar commented on nixos/modules/services/x11/desktop-managers/gnome.nix:
```diff
@@ -189,6 +189,12 @@ in
         };
       };
+
+      waylandOnly = lib.mkOption {
```
Might also want to clarify that it does not apply to flashback, which is always X11. But it should hopefully still be compatible if one uses both, as `gnome-flashback` creates its own session files.

Scrumplex reviewed (commented):
Changes LGTM. Is there a timeline to make this option the default at some point?

jtojnar reviewed (commented):
Nice, thanks.

jtojnar commented:
> Is there a timeline to make this option the default at some point?

We will follow upstream GNOME. See https://gitlab.gnome.org/GNOME/gnome-session/-/merge_requests/99 for that.

---

Also related: https://gitlab.gnome.org/GNOME/gnome-session/-/merge_requests/112

Scrumplex commented:
> We will follow upstream GNOME. See [gitlab.gnome.org/GNOME/gnome-session/-/merge_requests/99](https://gitlab.gnome.org/GNOME/gnome-session/-/merge_requests/99) for that.

Seems like this is more relevant: https://gitlab.gnome.org/GNOME/gdm/-/merge_requests/277

I guess once this one is merged, we can make it the default, as it would mark the deprecation of the X11 session

jtojnar commented:
Hmm, good catch. It might be better to expose that config option in GDM module instead of all the override shenanigans.

getchoo commented:
> It might be better to expose that config option in GDM module instead of all the override shenanigans.

I'm not a big fan of it either, but I do think there is some value here in letting users have a true "Xorg-less" desktop environment. Not sure if that's something in scope for this module, or modules in general though (maybe a global config option...?)

jtojnar commented:
> I'm not a big fan of it either, but I do think there is some value here in letting users have a true "Xorg-less" desktop environment.

Well, it won’t be “Xorg-less” because everything else still depends on it. And at that point, I would probably prefer the easier solution of disabling it in GDM.

And having a module that removes it everywhere would not be worth the substantial effort IMO.

Hopefully, it will be removed from the DE upstream in 49 or 50. But apps will still without a doubt continue to support X for many years.

MattSturgeon commented on nixos/modules/services/x11/desktop-managers/gnome.nix:
```diff
@@ -215,7 +225,9 @@ in
 
-      services.displayManager.sessionPackages = [ pkgs.gnome-session.sessions ];
+      services.displayManager.sessionPackages = [
+        (pkgs.gnome-session.override { x11Support = !cfg.waylandSessionOnly; }).sessions
```
The package is also used on another two lines below [L322](https://github.com/getchoo-contrib/nixpkgs/blob/b18b9492f0f0f6bbf3a0f5ea9dd79aece6d78836/nixos/modules/services/x11/desktop-managers/gnome.nix#L322) and [L365](https://github.com/getchoo-contrib/nixpkgs/blob/b18b9492f0f0f6bbf3a0f5ea9dd79aece6d78836/nixos/modules/services/x11/desktop-managers/gnome.nix#L365).

It seems likely that we could end up building/installing multiple versions of the package if we override one but not the others.

Should we bind the overridden package in a let-in block, and replace _all_ references to `pkgs.gnome-session` with it?

MattSturgeon reviewed (commented):
Looks good overall

getchoo committed:
nixos/gdm: add `xorg` option

This allows toggling the Xorg session, similar to the `wayland` option.
It's primarily intended to allow for using the Wayland session
exclusively

getchoo commented:
> And having a module that removes it everywhere would not be worth the substantial effort IMO.

Yeah thinking about that again, it'd be a total mess lol

> And at that point, I would probably prefer the easier solution of disabling it in GDM.

Implemented this instead. Let me know if I handled the Xsession wrapping stuff right!

jtojnar commented on nixos/doc/manual/release-notes/rl-2505.section.md:
```diff
@@ -555,6 +555,8 @@
 
 - `services.cloudflared` now uses a dynamic user, and its `user` and `group` options have been removed. If the user or group is still necessary, they can be created manually.
 
+- GNOME/GDM's X11 session can be disabled via the new [`services.xserver.displayManager.gdm.xorg`](#opt-services.xserver.displayManager.gdm.xorg) option.
```
It should be possible to use a shortcut:

```suggestion
- GNOME/GDM's X11 session can be disabled via the new [](#opt-services.xserver.displayManager.gdm.xorg) option.
```

MattSturgeon reviewed (commented):
I'm not really familiar with the gdm side of things, but just reviewing the nix code:

getchoo commented on nixos/modules/services/x11/display-managers/gdm.nix:
```diff
@@ -179,12 +186,18 @@ in
-        // lib.optionalAttrs (xSessionWrapper != null) {
+
+        (lib.mkIf cfg.gdm.xorg {
+          GDM_X_SERVER_EXTRA_ARGS = toString (lib.filter (arg: arg != "-terminate") cfg.xserverArgs);
```
This is not introduced by this PR

MattSturgeon reviewed (approved):
Looks good from my perspective, once formatted.

As I said before, I'm not qualified to sign off whether everything is correct from gdm's perspective. But I don't see any issues in the diff :slightly_smiling_face:

nyabinary commented:
Merge conflict with treewide format :3
